swarm-test/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+            .external_addresses()
+            .cloned()
+            .map(|r| r.addr)
+            .collect();
+
+        let dial_opts = DialOpts::peer_id(*other.local_peer_id())
+            .addresses(external_addresses)
+            .build();
+
+        self.dial(dial_opts).unwrap();
+
+        let mut dialer_done = false;
+        let mut listener_done = false;
+
+        loop {
+            match futures::future::select(self.next_swarm_event(), other.next_swarm_event()).await {
+                Either::Left((SwarmEvent::ConnectionEstablished { .. }, _)) => {
+                    dialer_done = true;
+                }
+                Either::Right((SwarmEvent::ConnectionEstablished { .. }, _)) => {
+                    listener_done = true;
+                }
+                Either::Left((other, _)) => {
+                    log::debug!("Ignoring event from dialer {:?}", other);
+                }
+                Either::Right((other, _)) => {
+                    log::debug!("Ignoring event from listener {:?}", other);
+                }
+            }
+
+            if dialer_done && listener_done {
+                return;
+            }
+        }
+    }
+
+    async fn dial_and_wait(&mut self, addr: Multiaddr) -> PeerId {
+        self.dial(addr.clone()).unwrap();
+
+        self.wait(|e| match e {
+            SwarmEvent::ConnectionEstablished {
+                endpoint, peer_id, ..
+            } => (endpoint.get_remote_address() == &addr).then_some(peer_id),
+            other => {
+                log::debug!("Ignoring event from dialer {:?}", other);
+                None
+            }
+        })
+        .await
+    }
+
+    async fn wait<E, P>(&mut self, predicate: P) -> E
+    where
+        P: Fn(SwarmEvent<<B as NetworkBehaviour>::OutEvent, THandlerErr<B>>) -> Option<E>,
+        P: Send,
+    {
+        loop {
+            let event = self.next_swarm_event().await;
+            if let Some(e) = predicate(event) {
+                break e;
+            }
+        }
+    }
+
+    async fn listen(&mut self) -> (Multiaddr, Multiaddr) {
+        let memory_addr_listener_id = self.listen_on(Protocol::Memory(0).into()).unwrap();
+
+        // block until we are actually listening
+        let memory_multiaddr = self
+            .wait(|e| match e {
+                SwarmEvent::NewListenAddr {
+                    address,
+                    listener_id,
+                } => (listener_id == memory_addr_listener_id).then_some(address),
+                other => {
+                    log::debug!(
+                        "Ignoring {:?} while waiting for listening to succeed",
+                        other
+                    );
+                    None
+                }
+            })
+            .await;
+
+        // Memory addresses are externally reachable because they all share the same memory-space.
+        self.add_external_address(memory_multiaddr.clone(), AddressScore::Infinite);
```

Comment:
:+1: good idea

Reply:
Thanks! I ended up doing it this way because it allowed to create the `connect` abstraction which can then rely on retrieving the addresses via the external addresses.